src/array/cpu/labor_pick.h

```diff
@@ -0,0 +1,347 @@
+#include <dmlc/omp.h>
+#include <dgl/runtime/parallel_for.h>
+#include <parallel_hashmap/phmap.h>
+#include <functional>
+#include <algorithm>
+#include <string>
+#include <vector>
+#include <numeric>
+#include <memory>
+#include <utility>
+#include <cmath>
+#include <iostream>
+#include <pcg_random.hpp>
+
+namespace dgl {
+namespace aten {
+namespace impl {
+
+// Template for picking non-zero values row-wise.
+template <typename IdxType, typename FloatType>
+std::pair<COOMatrix, FloatArray> CSRLaborPick(
+    CSRMatrix mat,
+    IdArray NIDs,
+    IdArray rows,
+    int64_t __num_picks,
+    FloatArray prob,
+    IdArray random_seed,
+    IdArray cnt,
+    int importance_sampling) {
+  using namespace aten;
+  const IdxType* indptr = static_cast<IdxType*>(mat.indptr->data);
+  const IdxType* indices = static_cast<IdxType*>(mat.indices->data);
+  const IdxType* data = CSRHasData(mat)? static_cast<IdxType*>(mat.data->data) : nullptr;
+  const IdxType* rows_data = static_cast<IdxType*>(rows->data);
+  const IdxType* nids = NIDs->shape[0] ? static_cast<IdxType*>(NIDs->data) : nullptr;
+  const auto num_rows = rows->shape[0];
+  const IdxType num_picks = __num_picks;
+  const auto& ctx = mat.indptr->ctx;
+
+  constexpr bool linear_c_convergence = false;
+
+  const bool weights = !IsNullArray(prob);
+  if (importance_sampling >= 0)
+    importance_sampling += weights;
+  auto A_arr = prob;
+  FloatType *A = static_cast<FloatType*>(A_arr->data);
+  constexpr double eps = 0.0001;
+  constexpr FloatType ONE = 1;
+
+  phmap::flat_hash_map<IdxType, FloatType> hop_map;
+
+  FloatArray cs_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* cs = static_cast<FloatType*>(cs_array->data);
+  FloatArray ds_array = NDArray::Empty({num_rows},
+                                        DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                        ctx);
+  FloatType* ds = static_cast<FloatType*>(ds_array->data);
+
+  IdxType max_d = 1;
+  IdxType hop_size = 0;
+  for (int64_t i = 0; i < num_rows; ++i) {
+    const IdxType rid = rows_data[i];
+    const auto act_d = indptr[rid + 1] - indptr[rid];
+    max_d = std::max(act_d, max_d);
+    double d = weights ? std::accumulate(A + indptr[rid], A + indptr[rid + 1], 0.0) : act_d;
+    cs[i] = num_picks / d;
+    ds[i] = d;
+    hop_size += act_d;
+  }
+
+  if (importance_sampling) {
+    FloatArray ps_array = NDArray::Empty({max_d + 1},
+                                          DGLDataType{kDGLFloat, 8*sizeof(FloatType), 1},
+                                          ctx);
+    FloatType* ps = static_cast<FloatType*>(ps_array->data);
+
+    double prev_ex_nodes = max_d * num_rows;
+
+    phmap::flat_hash_map<IdxType, FloatType> hop_map2;
+    for (int iters = 0; iters < importance_sampling || importance_sampling < 0; iters++) {
+      if (!weights || iters) {
```

Comment:
I can't quite map the logic of this if-block to the algorithm.  Could you describe what it is doing as a comment?  (Or you can give a reference to the paper).

Reply:
There is no reference in the paper. It is an implementation detail about when there are no weights, then the first c values can be computed in O(1) because all the edge weights are uniform. The c value is just k / d where k is fanout and d is the degree. But when there are weights, since the computed c values are incorrect, I have to skip some logic making use of it to compute the right c values below.

Reply:
The later iterations will have correct c values so they will take the if in that case.

Reply:
I also increase importance_sampling by 1 above so that the weighted case can do one more of this iteration to match the unweighted case.

Reply:
I will improve the overall logic and seperate things into functions so that the code becomes more readable.